File: backend/src/main/java/com/e_commerce/backend/feature_product/service/CategoryService.java
package com.e_commerce.backend.feature_product.service;

import com.e_commerce.backend.feature_product.model.CategoryEntity;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

public interface CategoryService {
    CategoryEntity createCategory(String name);
    List<CategoryEntity> getAllCategories();
    List<CategoryEntity> getActiveCategories();
    Page<CategoryEntity> getAllCategories(Pageable pageable);
    CategoryEntity getCategoryById(UUID id);
    CategoryEntity updateCategory(UUID id, String name);
    CategoryEntity hideCategory(UUID id);
    CategoryEntity unhideCategory(UUID id);
    void deleteCategory(UUID id);
}

File: frontend/src/app/core/services/category.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Category, PageResponse } from '../models/product.model';
import { ApiResponse } from '../models/api-response.model';

export interface CategoryRequest {
  name: string;
}

@Injectable({
  providedIn: 'root'
})
export class CategoryService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/categories`;

  getCategories(page: number = 0, size: number = 10): Observable<ApiResponse<PageResponse<Category>>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<ApiResponse<PageResponse<Category>>>(this.apiUrl, { params });
  }

  getAllCategories(): Observable<ApiResponse<Category[]>> {
    return this.http.get<ApiResponse<Category[]>>(`${this.apiUrl}/all`);
  }

  getCategoryById(id: string): Observable<ApiResponse<Category>> {
    return this.http.get<ApiResponse<Category>>(`${this.apiUrl}/${id}`);
  }

  createCategory(request: CategoryRequest): Observable<ApiResponse<Category>> {
    return this.http.post<ApiResponse<Category>>(this.apiUrl, request);
  }

  updateCategory(id: string, request: CategoryRequest): Observable<ApiResponse<Category>> {
    return this.http.put<ApiResponse<Category>>(`${this.apiUrl}/${id}`, request);
  }

  hideCategory(id: string): Observable<ApiResponse<Category>> {
    return this.http.patch<ApiResponse<Category>>(`${this.apiUrl}/${id}/hide`, {});
  }

  unhideCategory(id: string): Observable<ApiResponse<Category>> {
    return this.http.patch<ApiResponse<Category>>(`${this.apiUrl}/${id}/unhide`, {});
  }

  deleteCategory(id: string): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${id}`);
  }
}
